// unenv@2.0.0-rc.19/dist/runtime/node/fs.mjs
import promises from "node:fs/promises";
import { Dir, Dirent, FileReadStream, FileWriteStream, ReadStream, Stats, WriteStream } from "./internal/fs/classes.mjs";
import { _toUnixTimestamp, access, accessSync, appendFile, appendFileSync, chmod, chmodSync, chown, chownSync, close, closeSync, copyFile, copyFileSync, cp, cpSync, createReadStream, createWriteStream, exists, existsSync, fchmod, fchmodSync, fchown, fchownSync, fdatasync, fdatasyncSync, fstat, fstatSync, fsync, fsyncSync, ftruncate, ftruncateSync, futimes, futimesSync, glob, lchmod, globSync, lchmodSync, lchown, lchownSync, link, linkSync, lstat, lstatSync, lutimes, lutimesSync, mkdir, mkdirSync, mkdtemp, mkdtempSync, open, openAsBlob, openSync, opendir, opendirSync, read, readFile, readFileSync, readSync, readdir, readdirSync, readlink, readlinkSync, readv, readvSync, realpath, realpathSync, rename, renameSync, rm, rmSync, rmdir, rmdirSync, stat, statSync, statfs, statfsSync, symlink, symlinkSync, truncate, truncateSync, unlink, unlinkSync, unwatchFile, utimes, utimesSync, watch, watchFile, write, writeFile, writeFileSync, writeSync, writev, writevSync } from "./internal/fs/fs.mjs";
import * as constants from "./internal/fs/constants.mjs";
import { F_OK, R_OK, W_OK, X_OK } from "./internal/fs/constants.mjs";
export { F_OK, R_OK, W_OK, X_OK } from "./internal/fs/constants.mjs";
export { promises, constants };
export * from "./internal/fs/fs.mjs";
export * from "./internal/fs/classes.mjs";
export default {
	F_OK,
	R_OK,
	W_OK,
	X_OK,
	constants,
	promises,
	Dir,
	Dirent,
	FileReadStream,
	FileWriteStream,
	ReadStream,
	Stats,
	WriteStream,
	_toUnixTimestamp,
	access,
	accessSync,
	appendFile,
	appendFileSync,
	chmod,
	chmodSync,
	chown,
	chownSync,
	close,
	closeSync,
	copyFile,
	copyFileSync,
	cp,
	cpSync,
	createReadStream,
	createWriteStream,
	exists,
	existsSync,
	fchmod,
	fchmodSync,
	fchown,
	fchownSync,
	fdatasync,
	fdatasyncSync,
	fstat,
	fstatSync,
	fsync,
	fsyncSync,
	ftruncate,
	ftruncateSync,
	futimes,
	futimesSync,
	glob,
	lchmod,
	globSync,
	lchmodSync,
	lchown,
	lchownSync,
	link,
	linkSync,
	lstat,
	lstatSync,
	lutimes,
	lutimesSync,
	mkdir,
	mkdirSync,
	mkdtemp,
	mkdtempSync,
	open,
	openAsBlob,
	openSync,
	opendir,
	opendirSync,
	read,
	readFile,
	readFileSync,
	readSync,
	readdir,
	readdirSync,
	readlink,
	readlinkSync,
	readv,
	readvSync,
	realpath,
	realpathSync,
	rename,
	renameSync,
	rm,
	rmSync,
	rmdir,
	rmdirSync,
	stat,
	statSync,
	statfs,
	statfsSync,
	symlink,
	symlinkSync,
	truncate,
	truncateSync,
	unlink,
	unlinkSync,
	unwatchFile,
	utimes,
	utimesSync,
	watch,
	watchFile,
	write,
	writeFile,
	writeFileSync,
	writeSync,
	writev,
	writevSync
};
