// statsig-node/dist/ConfigEvaluation.js
"use strict";
exports.__esModule = true;
var ConfigEvaluation = /** @class */ (function () {
    function ConfigEvaluation(value, rule_id, secondary_exposures, json_value, explicit_parameters, config_delegate, fetch_from_server) {
        if (rule_id === void 0) { rule_id = ''; }
        if (secondary_exposures === void 0) { secondary_exposures = []; }
        if (json_value === void 0) { json_value = {}; }
        if (explicit_parameters === void 0) { explicit_parameters = null; }
        if (config_delegate === void 0) { config_delegate = null; }
        if (fetch_from_server === void 0) { fetch_from_server = false; }
        this.value = value;
        this.rule_id = rule_id;
        if (typeof json_value === 'boolean') {
            // handle legacy gate case
            this.json_value = {};
        }
        else {
            this.json_value = json_value;
        }
        this.secondary_exposures = secondary_exposures;
        this.undelegated_secondary_exposures = secondary_exposures;
        this.config_delegate = config_delegate;
        this.fetch_from_server = fetch_from_server;
        this.explicit_parameters = explicit_parameters;
        this.is_experiment_group = false;
    }
    ConfigEvaluation.prototype.withEvaluationDetails = function (evaulationDetails) {
        this.evaluation_details = evaulationDetails;
        return this;
    };
    ConfigEvaluation.prototype.setIsExperimentGroup = function (isExperimentGroup) {
        if (isExperimentGroup === void 0) { isExperimentGroup = false; }
        this.is_experiment_group = isExperimentGroup;
    };
    ConfigEvaluation.fetchFromServer = function () {
        return new ConfigEvaluation(false, '', [], {}, undefined, undefined, true);
    };
    return ConfigEvaluation;
}());
exports["default"] = ConfigEvaluation;
