// @vercel/examples-ui/dist/index.js
import _extends from "@swc/helpers/src/_extends.mjs";
import _object_without_properties_loose from "@swc/helpers/src/_object_without_properties_loose.mjs";
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import cn from 'clsx';
export { default as Layout } from './layout.js';
export { default as Text } from './text.js';
export { default as Input } from './input.js';
export { default as Button } from './button.js';
export { default as LoadingDots } from './loading-dots.js';
export { default as Link, A } from './link.js';
export { default as Code } from './code.js';
export { default as List } from './list.js';
export { default as Snippet } from './snippet.js';
const Noop = ({ children  })=>/*#__PURE__*/ _jsx(_Fragment, {
        children: children
    });
export const Page = (_param)=>/*#__PURE__*/ {
    var { children , className  } = _param, props = _object_without_properties_loose(_param, [
        "children",
        "className"
    ]);
    return _jsx("main", _extends({}, props, {
        className: cn('w-full max-w-3xl mx-auto py-16', className),
        children: children
    }));
};
export function getLayout(Component) {
    return Component.Layout || Noop;
}
