// @vercel/examples-ui/dist/layout.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import Head from 'next/head.js';
import Nav from './nav.js';
import Vercel from './icons/vercel.js';
const Layout = ({ title , description , path , deployButton , children ,  })=>{
    return /*#__PURE__*/ _jsxs("div", {
        className: "mx-auto h-screen flex flex-col",
        children: [
            /*#__PURE__*/ _jsxs(Head, {
                children: [
                    title && /*#__PURE__*/ _jsx("title", {
                        children: `${title} - Vercel Examples`
                    }),
                    description && /*#__PURE__*/ _jsx("meta", {
                        name: "description",
                        content: description
                    }),
                    /*#__PURE__*/ _jsx("link", {
                        rel: "icon",
                        href: "/favicon.ico"
                    })
                ]
            }),
            /*#__PURE__*/ _jsx(Nav, {
                path: path,
                deployButton: deployButton
            }),
            /*#__PURE__*/ _jsx("div", {
                className: "px-8 bg-accents-0",
                children: children
            }),
            /*#__PURE__*/ _jsxs("footer", {
                className: "py-10 w-full mt-auto border-t flex items-center justify-center bg-accents-1 z-20",
                children: [
                    /*#__PURE__*/ _jsx("span", {
                        className: "text-primary",
                        children: "Created by"
                    }),
                    /*#__PURE__*/ _jsx("a", {
                        href: "https://vercel.com",
                        "aria-label": "Vercel.com Link",
                        target: "_blank",
                        rel: "noreferrer",
                        className: "text-black",
                        children: /*#__PURE__*/ _jsx(Vercel, {
                            className: "inline-block h-6 ml-3 text-primary",
                            alt: "Vercel.com Logo"
                        })
                    })
                ]
            })
        ]
    });
};
export default Layout;
