// @vercel/examples-ui/dist/nav.js
import _extends from "@swc/helpers/src/_extends.mjs";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import Link from './link.js';
import Button from './button.js';
import DeployButton from './deploy-button.js';
const REPO_URL = 'https://github.com/vercel/examples/tree/main';
export default function Nav({ path , deployButton  }) {
    const displayPath = [
        'Vercel Examples'
    ].concat((path == null ? void 0 : path.split('/').filter(Boolean)) || []).join(' / ');
    const repositoryUrl = (deployButton == null ? void 0 : deployButton.repositoryUrl) || `${REPO_URL}/${path}`;
    return /*#__PURE__*/ _jsx("nav", {
        className: "border-b border-gray-200 py-5 relative z-20 bg-background shadow-[0_0_15px_0_rgb(0,0,0,0.1)]",
        children: /*#__PURE__*/ _jsxs("div", {
            className: "flex items-center lg:px-6 px-8 mx-auto max-w-7xl px-14",
            children: [
                /*#__PURE__*/ _jsxs("div", {
                    className: "flex flex-row items-center",
                    children: [
                        /*#__PURE__*/ _jsx(Link, {
                            href: "/",
                            children: /*#__PURE__*/ _jsx("span", {
                                children: /*#__PURE__*/ _jsxs("svg", {
                                    height: "26",
                                    viewBox: "0 0 75 65",
                                    fill: "#000",
                                    children: [
                                        /*#__PURE__*/ _jsx("title", {
                                            children: "Vercel Logo"
                                        }),
                                        /*#__PURE__*/ _jsx("path", {
                                            d: "M37.59.25l36.95 64H.64l36.95-64z"
                                        })
                                    ]
                                })
                            })
                        }),
                        /*#__PURE__*/ _jsxs("ul", {
                            className: "flex items-center content-center",
                            children: [
                                /*#__PURE__*/ _jsx("li", {
                                    className: "ml-2 text-gray-200",
                                    children: /*#__PURE__*/ _jsx("svg", {
                                        viewBox: "0 0 24 24",
                                        width: "32",
                                        height: "32",
                                        stroke: "currentColor",
                                        strokeWidth: "1",
                                        strokeLinecap: "round",
                                        strokeLinejoin: "round",
                                        fill: "none",
                                        shapeRendering: "geometricPrecision",
                                        children: /*#__PURE__*/ _jsx("path", {
                                            d: "M16.88 3.549L7.12 20.451"
                                        })
                                    })
                                }),
                                /*#__PURE__*/ _jsx("li", {
                                    className: "font-medium",
                                    style: {
                                        letterSpacing: '.01px'
                                    },
                                    children: /*#__PURE__*/ _jsx(Link, {
                                        href: repositoryUrl,
                                        className: "text-accents-6 no-underline transition-colors duration-200 hover:text-accents-8 cursor-pointer",
                                        target: "_blank",
                                        rel: "noreferrer",
                                        children: displayPath
                                    })
                                })
                            ]
                        })
                    ]
                }),
                /*#__PURE__*/ _jsx("div", {
                    className: "flex-1 justify-end hidden md:flex",
                    children: /*#__PURE__*/ _jsxs("nav", {
                        className: "flex-row inline-flex items-center",
                        children: [
                            /*#__PURE__*/ _jsx("span", {
                                className: "ml-2 h-full flex items-center cursor-not-allowed text-accents-5",
                                children: /*#__PURE__*/ _jsx(Button, {
                                    variant: "ghost",
                                    Component: "a",
                                    href: "https://github.com/vercel/examples/tree/main",
                                    target: "_blank",
                                    rel: "noreferrer",
                                    children: "More Examples →"
                                })
                            }),
                            /*#__PURE__*/ _jsx("span", {
                                className: "ml-2 h-full flex items-center cursor-not-allowed text-accents-5",
                                children: /*#__PURE__*/ _jsx(DeployButton, _extends({}, deployButton, {
                                    repositoryUrl: repositoryUrl,
                                    projectName: (deployButton == null ? void 0 : deployButton.projectName) || path,
                                    repositoryName: (deployButton == null ? void 0 : deployButton.repositoryName) || path
                                }))
                            })
                        ]
                    })
                })
            ]
        })
    });
}
